smoke-tests/tests/test_functions.py

```diff
@@ -0,0 +1,176 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import dataclasses
+import os
+import pathlib
+import subprocess
+
+import pytest
+import requests
+from urllib3.util.retry import Retry
+
+# This should be the parent directory of `tests/`
+CWD = pathlib.Path.cwd()
+
+
+@dataclasses.dataclass
+class CloudFunction:
+    """Hold information for a Google Cloud Function."""
+
+    region: str
+    target: str
+    cwd: pathlib.Path
+    port: int = dataclasses.field(repr=False)
+    protocol: str = dataclasses.field(repr=False, default="http")
+    host: str = dataclasses.field(repr=False, default="localhost")
+    url: str = dataclasses.field(init=False)
+
+    def __post_init__(self):
+        """Create the value for the url field dynamically."""
+        self.url = f"{self.protocol}://{self.host}:{self.port}"
+
+
+@pytest.fixture(name="client_us", scope="session")
+def fixture_client_us() -> CloudFunction:
+    """Return a Function instance for a client in the US region."""
+
+    return CloudFunction(
+        region="US", target="run_geo_smoke_test", cwd=CWD / "client", port=8844
+    )
+
+
+@pytest.fixture(name="client_gb", scope="session")
+def fixture_client_gb() -> CloudFunction:
+    """Return a Function instance for a client in the GB region."""
+
+    return CloudFunction(
+        region="GB", target="run_geo_smoke_test", cwd=CWD / "client", port=8845
+    )
+
+
+@pytest.fixture(name="client_ch", scope="session")
+def fixture_client_ch() -> CloudFunction:
+    """Return a Function instance for a client in the CH region."""
+
+    return CloudFunction(
+        region="CH", target="run_geo_smoke_test", cwd=CWD / "client", port=8846
+    )
+
+
+@pytest.fixture(name="runner", scope="session")
+def fixture_runner() -> CloudFunction:
+    """Return a Function instance for the runner."""
+    return CloudFunction(
+        region="US", target="run_geo_smoke_tests", cwd=CWD / "runner", port=8847
+    )
+
+
+@pytest.fixture(name="client_functions", scope="session", autouse=True)
+def fixture_client_functions(
+    client_us: CloudFunction,
+    client_gb: CloudFunction,
+    client_ch: CloudFunction,
+):
+    """Start and terminate new subprocesses for the functions."""
+
+    p_clients = [
+        subprocess.Popen(
+            [
+                "functions-framework",
+                "--target",
+                client.target,
+                "--port",
+                str(client.port),
```

Comment:
Curious as to why you did it this was vs importing ```functions-framework```

Reply:
It's the documented method of running integration tests. 📋 

See https://cloud.google.com/functions/docs/testing/test-http#integration_tests

Reply:
I pushed a commit to document that in code.